src/aspire/denoise/denoise_cov2d.py

```diff
@@ -0,0 +1,98 @@
+
+import logging
+import numpy as np
+
+from aspire.denoise import Denoise
+from aspire.utils.filters import RadialCTFFilter
+from aspire.utils.blk_diag_func import radial_filter2fb_mat
+from aspire.utils.blk_diag_func import blk_diag_partition
+from aspire.utils.blk_diag_func import blk_diag_eye
+from aspire.basis.ffb_2d import FFBBasis2D
+from aspire.estimation.covar2d import RotCov2D
+from aspire.utils.matrix import anorm
+
+logger = logging.getLogger(__name__)
+
+
+class DenoiseCov2D(Denoise):
+    """
+    Define a derived class for denoising 2D images using Cov2D method
+    """
+    def __init__(self, src, math_basis, var_noise, ctf_info):
+        """
+        Initialize an object for denoising 2D images using Cov2D method
+
+        :param src: The source object of 2D images with metadata
+        :param math_basis: The basis method to expand 2D images
+        :param var_noise: The estimated variance of noise
+        :param ctf_info: Whether the CTF filters are included
+        """
+        super().__init__(src)
+        self.var_noise = var_noise
+        if not isinstance(math_basis, FFBBasis2D):
+            raise NotImplementedError('Currently only Fast FB method is supported')
+        self.coeffs_noise = None
+        self.coeffs_estim = None
+        self.ctf_idx = None
+        self.ctf_fb = None
+
+        # Assign the CTF information and index for each image
+        if ctf_info and not (src.filters is None):
+            logger.info(f'Convert non radical CTF filters to radical ones')
+            uniq_rad_filters = []
+            uniq_ctf_filters = list(set(src.filters))
+            for f in uniq_ctf_filters:
+                # convert non radical CTF filters to radial ones
+                defocus = np.sqrt(f.defocus_u*f.defocus_u + f.defocus_v*f.defocus_v)
+                uniq_rad_filters.append(RadialCTFFilter(
+                    f.pixel_size, f.voltage, defocus=defocus, Cs=f.Cs, alpha=f.alpha))
+            self.ctf_idx = np.array([uniq_ctf_filters.index(f) for f in src.filters])
+            # Evaluate CTF in the FFB basis
+            self.ctf_fb = [radial_filter2fb_mat(f.evaluate, math_basis)
+                           for f in uniq_rad_filters]
+        else:
+            logger.info(f'CTF filters are not included in Cov2D denoising.')
+            f = RadialCTFFilter(1.0, 200, defocus=1.5e4, Cs=2.0, alpha=0.1)
+            # set CTF filters to an identity filter
+            self.ctf_idx = np.zeros(self.nimg, dtype=int)
+            self.ctf_fb = [blk_diag_eye(blk_diag_partition(radial_filter2fb_mat(f.evaluate, math_basis)))]
```

Comment:
Shouldn't this be done in each of the lower-end functions (e.g., `get_mean`, `get_covar`, `get_cwf_coeffs`)? Right now, they have to implement their own special case for `ctf_fb == None` (see below). Wouldn't it be easier to just set both of these to `None` here and handle this in the other functions. Otherwise we have the same code in two places (sort of).

Reply:
We have the additional case of simulation data using the clean images. We have arranged it in this way, if ctf_fb or ctf_idx is none, we should call it for cov2d of clean images of simulation data; For experiment data targeted for this class, the ctf_fb or ctf_idx can not be None, either read from star file or set to identity.

Reply:
More explanation are added in the commit 059f18f.

Reply:
I still don't understand why not specifying a CTF means that images necessarily have to be clean. There are four scenarios:
– CTF and noise (normal for experimental images)
– No CTF and noise (if CTF is not know, or if we have some sort of CTF correction already)
– CTF and no noise (clean, but filtered)
– No CTF and no noise (clean, but unfiltered)
We need to be able to handle all of these. Just because there is no CTF specified, we shouldn't assume that the images are clean.

Reply:
The Cov2D module has all four options. I tend to limit this  derived class for denoising experimental data. We do need to include all four cases if we want to include the simulation data.

Reply:
Changed to use the identity filter when the CTF filters are not specified in commit b6a96b. The `coeffs` and `noise_var` will determine whether the images are clean or not.